crates/schema-inference/src/schema.rs

```diff
@@ -0,0 +1,118 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use doc::inference::Shape;
+use json::schema::types::{self, Set};
+use schemars::gen::SchemaGenerator;
+use schemars::schema::{InstanceType, RootSchema, Schema, SchemaObject, SingleOrVec};
+
+use crate::shape;
+
+#[derive(Debug, Default)]
+pub struct SchemaBuilder {
+    shape: Shape,
+}
+
+impl SchemaBuilder {
+    pub fn new(shape: Shape) -> Self {
+        Self { shape }
+    }
+
+    pub fn root_schema(&self) -> RootSchema {
+        RootSchema {
+            schema: to_schema(&self.shape).into_object(),
+            meta_schema: SchemaGenerator::default().settings().meta_schema.clone(),
+            ..Default::default()
+        }
+    }
+
+    pub fn merge(self, other: Self) -> Self {
+        Self {
+            shape: shape::merge(self.shape, other.shape),
+        }
+    }
+}
+
+pub fn to_schema(shape: &Shape) -> Schema {
+    let mut schema_object = SchemaObject {
+        instance_type: Some(shape_type_to_schema_type(shape.type_)),
+        ..Default::default()
+    };
+
+    if shape.type_.overlaps(types::OBJECT) {
+        let mut prop_schemas = BTreeMap::new();
+        let mut required = BTreeSet::new();
+        for obj_prop in shape.object.properties.iter() {
+            prop_schemas.insert(obj_prop.name.clone(), to_schema(&obj_prop.shape));
+            if obj_prop.is_required {
+                required.insert(obj_prop.name.clone());
+            }
+        }
+        let object = &mut schema_object.object();
+        object.properties = prop_schemas;
+        object.required = required;
+    }
+
+    if shape.type_.overlaps(types::ARRAY) {
+        let mut array_items = Vec::new();
+        for item in shape.array.tuple.iter() {
+            array_items.push(to_schema(item));
```

Comment:
Going along with the previous `tuple` comment, I think it seems best if we don't make any assumptions about the length of arrays when we're inferring schemas. Given an input array of `["foo", 3, true]`, I'm thinking that it would be better to produce a schema like `{"type": "array", "items": {"type": ["string", "integer", "boolean"]}}`, rather than having `items` be an array of schemas.